File: apps/parser/src/integrations/lastfm.ts
import { URLSearchParams } from 'url';

import { Injectable, OnApplicationBootstrap, OnModuleInit } from '@nestjs/common';
import { PrismaService } from '@tsuwari/prisma';

import { app } from '../index.js';

export interface LastFMDbData {
  username: string;
}

interface ILastfmResponse {
  error?: number;
  message?: string;
  recenttracks?: {
    track?: Array<{
      artist: {
        '#text': string;
      };
      '@attr'?: {
        nowplaying: boolean;
      };
      album?: {
        '#text': string;
      };
      name: string;
    }>;
  };
}

@Injectable()
export class LastFmIntegration implements OnModuleInit {
  #apiKey: string;

  constructor(private readonly prisma: PrismaService) { }

  async onModuleInit() {
    const service = await this.prisma.integration.findFirst({
      where: {
        service: 'LASTFM',
      },
    });

    if (service?.apiKey) {
      this.#apiKey = service.apiKey;
    }

    return this;
  }

  async fetchSong(userName: string) {
    const params = {
      method: 'user.getrecenttracks',
      user: userName,
      api_key: this.#apiKey,
      format: 'json',
      limit: '1',
    } as Record<string, string>;

    const request = await fetch(`http://ws.audioscrobbler.com/2.0?${new URLSearchParams(params)}`, {
      method: 'GET',
    });

    const response = await request.json() as ILastfmResponse;

    const song = response.recenttracks?.track?.find((t) => t['@attr']?.nowplaying);
    if (!song) return null;

    return `${song.artist['#text']} — ${song.name}`;
  }
}

File: apps/parser/internal/integrations/lastfm/lastfm.go
package lastfm

import (
	"fmt"

	model "github.com/satont/tsuwari/libs/gomodels"

	lfm "github.com/shkh/lastfm-go/lastfm"
)

type LastFm struct {
	integration *model.ChannelsIntegrations
}

func New(integration *model.ChannelsIntegrations) *LastFm {
	if integration == nil || !integration.APIKey.Valid || !integration.Integration.APIKey.Valid ||
		!integration.Integration.ClientSecret.Valid {
		return nil
	}

	service := LastFm{
		integration: integration,
	}

	return &service
}

func (c *LastFm) GetTrack() *string {
	api := lfm.New(
		c.integration.Integration.APIKey.String,
		c.integration.Integration.ClientSecret.String,
	)
	api.SetSession(c.integration.APIKey.String)

	user, err := api.User.GetInfo(map[string]interface{}{})
	if err != nil {
		return nil
	}

	tracks, err := api.User.GetRecentTracks(map[string]interface{}{
		"limit": "1",
		"user":  user.Name,
	})

	if err != nil || len(tracks.Tracks) == 0 || tracks.Tracks[0].NowPlaying != "true" {
		return nil
	}

	track := tracks.Tracks[0]

	response := fmt.Sprintf("%s — %s", track.Artist.Name, track.Name)

	return &response
}
